### ISSUE TITLE
The vanilla PDO example is all wrong, undermining the whole idea behind the wrapper. NOT AN ISSUE, JUST AN OPEN DISCUSSION

### ISSUE THREAD

**Issue by colshrapnel** [rawsrc/PDOPlusPlus#2]
The **real** vanilla PDO example should be

```php
$sql = 'INSERT INTO t_video
(video_title, video_support, video_multilingual, video_chapter, video_year, video_summary, video_stock)
VALUES (:video_title, :video_support, :video_multilingual, :video_chapter, :video_year, :video_summary, :video_stock)';
$pdo->prepare($sql)->execute($film);
$id = $pdo->lastInsertId();
```

which is actually *much* simpler than yours

```php
$in  = $ppp->injectorInByVal();
$sql = <<<sql
INSERT INTO t_video (
    video_title, video_support, video_multilingual, video_chapter, video_year, video_summary, video_stock
) VALUES (
   {$in($film['title'])}, {$in($film['support'])}, {$in($film['multilingual'], 'bool')}, 
   {$in($film['chapter'], 'int')}, {$in($film['year'], 'int')}, {$in($film['summary'])}, 
   {$in($film['stock'], 'int')}
);
sql;
$id = $ppp->insert($sql);
```

**Comment by colshrapnel** [comment]
I don't see any issues with sub-queries, to be honest. However, you are welcome to provide an example where PDO has limitations but your class has not. After all, it will be to your benefit.

Thank you for your advise. As you can see, I did read the PDO manual very well, so I was able to optimize your artificially blown up PDO code. Now your turn - show me the case where your class will beat vanilla PDO. We will see.    
After all, a honest comparison would be to your benefit.

**Comment by rawsrc** [comment]
No problem just with your code : 
```php
$sql = 'INSERT INTO t_video
(video_title, video_support, video_multilingual, video_chapter, video_year, video_summary, video_stock)
VALUES (:video_title, :video_support, :video_multilingual, :video_chapter, :video_year, :video_title, :video_stock)';
$pdo->prepare($sql)->execute($film);
```
Just use the video title also for the video summary for example.

Let's see another example : 
Try to have dynamic `LIMIT` clause using PDO:
```php
$sql = 'SELECT * FROM t_video LIMIT :start, :length';
$data = $pdo->prepare($sql)->execute([$start, $length])->fetchAll();
```
Using PDOPlusPlus, it's clearly possible:
```php
$ppp = new PPP();
$sql = "SELECT * FROM t_video LIMIT {$ppp($start, 'int')}, {$ppp($length, 'int')}"
$data = $ppp->select($sql); // no error, you get the dataset
```
I highly prefer to use explicit casting on binding than let the engine to treat all as strings, it's my opinion.
Anyway, if you're not convinced by this tool, it's not a problem, believe me :-)
Where's your code based on `->bindParam()` ?  ;-)

**Comment by colshrapnel** [comment]
Well, for a simple wrapper for select queries your class is way too overengineered, as you need only three lines to create such a wrapper, not three thousand.    
Ironically, it's _still_ inferior to PDO. 

```php
$sql = 'SELECT * FROM t_video LIMIT ?, ?';
$stmt-> $pdo->prepare($sql);
$stmt->execute([$start, $length]);
$data = $stmt->fetchAll(PDO:FETCH_UNIQUE); // already indexed by id 
```
Vanilla PDO can return the select result in dozens different formats while your crippled wrapper supports only _one_ :)

```php
$sql = "SELECT * FROM t_video LIMIT {$ppp($start, 'int')}, {$ppp($length, 'int')}"
$data = [];
foreach ($ppp->select($sql) as $row) {  // lol
    $data[$row['id']] = $row;
}
```

**Comment by rawsrc** [comment]
because you use dozens different formats in you code ? If you desire, i can add your dozen different formats in the next release :-)
I just note that you have to alternate between question marks and tags to be able to use it through PDO.

**Comment by colshrapnel** [comment]
Yes, I definitely have to, so I fixed your PDO code above that mixed them. 

To be honest, there are indeed edge cases for PDO but surprisingly, your wrapper doesn't address them at all, although it has all the means to. For example, the infamous IN() clause which can be handled naively with your wrapper.

**Comment by rawsrc** [comment]
i will let the dev to define the desired fetch mode in the next release and probably add also a way to manage a scrollable cursor.
Thanks a lot guy for the idea, i appreciate.

**Comment by rawsrc** [comment]
i released a new version of `PDOPlusPlus` (v3.2.0) that has some improvements regarding our discussion and is fully compliant with the previous.
Here's the new code : 
```php
$sql  = "SELECT * FROM t_video LIMIT {$ppp($start, 'int')}, {$ppp($length, 'int')}";
$data = $ppp->selectStmt($sql)->fetchAll(PDO:FETCH_UNIQUE);
```

**Comment by rawsrc** [comment]
Tanks again **colshrapnel** 
I close this ask for enhancement ;-)
